File: code/typescript-for-csharp-java/compare/L04Variance.java
// compare/L04Variance.java
import java.util.List;

public class L04Variance {
    record Instrument(String name) {}

    public static void main(String[] args) {
        String[] strings = { "E", "A" };
        Object[] objects = strings; // arrays are covariant in Java too
        try {
            objects[0] = 440;
        } catch (ArrayStoreException e) {
            System.out.println("objects[0] = 440: " + e.getClass().getSimpleName());
        }
        List<? extends Object> producer = List.of("E", "A"); // use-site variance: a wildcard
        System.out.println("producer.get(0): " + producer.get(0));
    }
}

File: code/typescript-for-csharp-java/compare/l04_variance.cs
// compare/l04_variance.cs
using System.Collections.Generic;

Guitar[] guitars = [new Guitar("guitar")];
Instrument[] instruments = guitars; // arrays are covariant in C# too
try { instruments[0] = new Instrument("piano"); }
catch (ArrayTypeMismatchException e) { Console.WriteLine($"instruments[0] = piano: {e.GetType().Name}"); }

IEnumerable<Instrument> sequence = new List<Guitar> { new("guitar") }; // IEnumerable<out T>
Action<Instrument> play = i => Console.WriteLine($"playing {i.Name}");
Action<Guitar> tune = play; // Action<in T>
tune(new Guitar("guitar"));
Console.WriteLine($"sequence: {string.Join(", ", sequence)}");

record Instrument(string Name);
record Guitar(string Name) : Instrument(Name);
